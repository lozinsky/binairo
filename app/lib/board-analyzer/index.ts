export * from './analyze-board-by-kind';
export * from './analyze-board';
