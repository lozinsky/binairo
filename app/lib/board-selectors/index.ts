export * from './select-filled-imbalanced-line';
export * from './select-line-that-can-be-equal-to-another';
export * from './select-line-that-includes-another';
export * from './select-line-that-is-equal-to-others';
export * from './select-middle-cell-between-two-identical';
export * from './select-next-cell-after-two-identical';
export * from './select-three-or-more-identical-sequential-cells';
export * from './select-unfilled-balanced-line';
