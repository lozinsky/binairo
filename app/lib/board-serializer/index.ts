export * from './board-serializer-collection';
