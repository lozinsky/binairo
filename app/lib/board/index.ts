export * from './board-cell';
export * from './board-line';
export * from './board';
