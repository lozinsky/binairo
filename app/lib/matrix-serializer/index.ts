export * from './matrix-selection-serializer-collection';
