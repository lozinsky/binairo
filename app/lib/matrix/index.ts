export * from './matrix-line';
export * from './matrix-rotation';
export * from './matrix-selection';
export * from './matrix';
