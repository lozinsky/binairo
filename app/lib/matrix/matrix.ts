import type { MatrixLine, MatrixLineCell } from './matrix-line';

export interface Matrix<L extends MatrixLine<unknown>> {
  at(index: number): L | undefined;
  entries(): Iterable<[number, L]>;
}

export type MatrixCell<M> = M extends Matrix<infer L> ? MatrixLineCell<L> : never;
