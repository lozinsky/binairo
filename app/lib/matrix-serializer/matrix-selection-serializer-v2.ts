import { isMatrixSelectionPosition, MatrixSelection, type MatrixSelectionPosition } from '~/lib/matrix';

import {
  type MatrixSelectionSerializer,
  MatrixSelectionSerializerDeserializeError,
} from './matrix-selection-serializer';

export class MatrixSelectionSerializerV2 implements MatrixSelectionSerializer {
  deserialize(data: string): MatrixSelection {
    if (data === '') {
      throw new MatrixSelectionSerializerDeserializeError('Invalid matrix selection length');
    }

    const positions: MatrixSelectionPosition[] = [];
    const coords = data.split('-');

    for (let index = 0; index < coords.length; index += 2) {
      const position = {
        x: Number(coords[index]),
        y: Number(coords[index + 1]),
      };

      if (!isMatrixSelectionPosition(position)) {
        throw new MatrixSelectionSerializerDeserializeError('Invalid matrix selection position');
      }

      positions.push(position);
    }

    return new MatrixSelection(positions);
  }

  serialize(selection: MatrixSelection): string {
    return selection
      .valueOf()
      .flatMap((position) => [position.x, position.y])
      .join('-');
  }
}
