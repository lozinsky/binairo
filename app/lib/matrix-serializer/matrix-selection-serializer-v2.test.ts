import { expect, test } from 'vitest';

import { MatrixSelection } from '~/lib/matrix';

import { MatrixSelectionSerializerV2 } from './matrix-selection-serializer-v2';

test.each(['', '-0-0-0-0-0-1-1-1', '0-0-0-0-0-1-1'])('throws on invalid data', (data) => {
  expect(() => new MatrixSelectionSerializerV2().deserialize(data)).toThrowErrorMatchingSnapshot();
});

test.each(['0-0-0-0-0-1-1-1', '0-0-0-0-0-1-1-1-0-2-2-2', '1-1-10-10-100-100-1000-1000-10000-10000'])(
  'returns matrix selection from string',
  (data) => {
    expect(new MatrixSelectionSerializerV2().deserialize(data)).toMatchSnapshot();
  },
);

test.each([
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
  ]),
  new MatrixSelection([
    { x: 0, y: 0 },
    { x: 0, y: 0 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
    { x: 0, y: 2 },
    { x: 2, y: 2 },
  ]),
  new MatrixSelection([
    { x: 1, y: 1 },
    { x: 10, y: 10 },
    { x: 100, y: 100 },
    { x: 1000, y: 1000 },
    { x: 10000, y: 10000 },
  ]),
])('returns matrix selection string', (selection) => {
  expect(new MatrixSelectionSerializerV2().serialize(selection)).toMatchSnapshot();
});
