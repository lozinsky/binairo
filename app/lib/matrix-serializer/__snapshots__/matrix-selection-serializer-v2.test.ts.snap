// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns matrix selection from string 1`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 1,
  },
  {
    "x": 1,
    "y": 1,
  },
]
`;

exports[`returns matrix selection from string 2`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 0,
    "y": 1,
  },
  {
    "x": 1,
    "y": 1,
  },
  {
    "x": 0,
    "y": 2,
  },
  {
    "x": 2,
    "y": 2,
  },
]
`;

exports[`returns matrix selection from string 3`] = `
[
  {
    "x": 1,
    "y": 1,
  },
  {
    "x": 10,
    "y": 10,
  },
  {
    "x": 100,
    "y": 100,
  },
  {
    "x": 1000,
    "y": 1000,
  },
  {
    "x": 10000,
    "y": 10000,
  },
]
`;

exports[`returns matrix selection string 1`] = `"0-0-0-0-0-1-1-1"`;

exports[`returns matrix selection string 2`] = `"0-0-0-0-0-1-1-1-0-2-2-2"`;

exports[`returns matrix selection string 3`] = `"1-1-10-10-100-100-1000-1000-10000-10000"`;

exports[`throws on invalid data 1`] = `[MatrixSelectionSerializerDeserializeError: Invalid matrix selection length]`;

exports[`throws on invalid data 2`] = `[MatrixSelectionSerializerDeserializeError: Invalid matrix selection position]`;

exports[`throws on invalid data 3`] = `[MatrixSelectionSerializerDeserializeError: Invalid matrix selection position]`;
